Make plantlike drawtype more fun.

sofar opened the pull request:
Adds several new ways that the plantlike drawtype mesh can be changed.

This requires paramtype2 = "meshoptions" to be set in the node
definition. The drawtype for these nodes should be "plantlike".

These modifications are all done using param2. This field is now
a complex bitfield that allows some or more of the combinations to
be chosen, and the mesh draw code will choose the options based as
needed for each plantlike node.

bit layout:
bits 0, 1 and 2 (values 0x1 through 0x7) are for choosing the plant
mesh shape:
  0 - ordinary plantlike plant ("x" shaped)
  1 - ordinary plant, but rotated 45 degrees ("+" shaped)
  2 - a plant with 3 faces ("*" shaped)
  3 - a plant with 4 faces ("#" shaped)
  4 through 7 are unused and reserved for future mesh shapes.

bit 3 (0x8) causes the plant to be randomly offset in the x,z
plane. The plant should fall within the 1x1x1 nodebox if regularly
sized.

bit 4 (0x10) causes the plant mesh to grow by sqrt(2), and will cause
the plant mesh to fill out 1x1x1, and appear slightly larger. Texture
makers will want to make their plant texture 23x16 pixels to have the
best visual fit in 1x1x1 size.

bit 5 (0x20) through bit 7 (0x80) are unused and reserved for
future use.

!(https://youtu.be/qWuI664krsI)
- [X] - CPT2_MESHOPTIONS
- [X] - don't stretch pixels
- [X] - Reorganize bitfield
- [x] - Docs

sofar committed:
Make plantlike drawtype more fun.

Adds several new ways that the plantlike drawtype mesh can be changed.

This requires paramtype2 = "meshoptions" to be set in the node
definition. The drawtype for these nodes should be "plantlike".

These modifications are all done using param2. This field is now
a complex bitfield that allows some or more of the combinations to
be chosen, and the mesh draw code will choose the options based as
neeeded for each plantlike node.

bit layout:
bits 0, 1 and 2 (values 0x1 through 0x7) are for choosing the plant
mesh shape:
  0 - ordinary plantlike plant ("x" shaped)
  1 - ordinary plant, but rotated 45 degrees ("+" shaped)
  2 - a plant with 3 faces ("*" shaped)
  3 - a plant with 4 faces ("#" shaped)
  4 - a plant with 4 faces ("#" shaped, leaning outwards)
  5 through 7 are unused and reserved for future mesh shapes.

bit 3 (0x8) causes the plant to be randomly offset in the x,z
plane. The plant should fall within the 1x1x1 nodebox if regularly
sized.

bit 4 (0x10) causes the plant mesh to grow by sqrt(2), and will cause
the plant mesh to fill out 1x1x1, and appear slightly larger. Texture
makers will want to make their plant texture 23x16 pixels to have the
best visual fit in 1x1x1 size.

bit 5 (0x20) causes each face of the plant to have a slight negative
Y offset in position, descending up to 0.125 downwards into the node
below. Because this is per face, this causes the plant model to be
less symmetric.

bit 6 (0x40) through bit 7 (0x80) are unused and reserved for
future use.

!(https://youtu.be/qWuI664krsI)

est31 commented:
You should add its own param2type for that. cheking for `CPT2_NONE` is bad quality. Its like adding your .doc files to `/`.

est31 commented:
For how to do that, see commit b0523adc5f673208b5414cf49c6640e4fc148b87 for example.

est31 commented:
This way the diff of the commit linked above reads much nicer (to filter out the whitespace cleanup):
https://github.com/minetest/minetest/commit/b0523adc5f673208b5414cf49c6640e4fc148b87?w=1

Its a fun idea.

sofar commented:
>You should add its own param2type for that. cheking for `CPT2_NONE` is
>bad quality. Its like adding your .doc files to `/`.

That's why it's marked wip :)

Yes, the bitfield usage and layout isn't optimal either, but I'd rather get some early feedback on the concept of using param2 in this matter.

tenplus1 commented:
Is there any way to make a plantlike node for underwater use that doesnt create an air bubble around it ?

kilbith commented:
Can it be complemented with https://github.com/minetest/minetest/pull/3446 ?

@celeron55 likes it by the way.

celeron55 commented:
This might indeed be the best use for param2 of plants. Of course, if this is taken into use, then param2 cannot be used for anything else.

I would expect some farming or whatever mods to be already using param2 for other custom stuff. Maybe there should also be a way to set a static value to be used as these bits for a given node in those cases so that they don't completely miss out on this?

Not sure about the exact functions of the bits though. I think none of them should cancel each other out at least so that we can get the most variation out of the least amount of bits.

I would just randomize all the bits for natural plants to make them look as wild as possible. 8)

ObaniGemini commented:
This looked weird to me on a screenshot but now I see it in the video it looks amazing :)

paramat commented:
This is better than the fine (degrees) rotation for plantlike, which i think should not have been added.

I like the different orientations, the hash version and the one with 6-fold symmetry.
Still not keen on randomised placement, as mapgen never places plants on every square of a grid, the already randomised placement disguises the central placement within a node. When plants are placed in every square of a grid this is usually a crop which looks better aligned anyway.

C1ffisme commented:
Nice ideas... I bet I know where you got that hash idea from, though.

sofar commented:
@kilbith 

> Can it be complemented with #3446 ?

It already is. bit 0x1 enables that behavior.

sofar commented:
@C1ffisme 

I had manually made a mesh like that for my ```crops``` mod, yes.

sofar commented:
@tenplus1 potentially, yes, but I'll have to look at the water face drawing parts first, and it may require that a node is marked as a liquid as well

sofar commented:
As  for the bitfield, I've got these right now:

General modifiers:
- sizer
- randomize placement
- angle changer

Mesh choices:
- normal
- 3 faced
- hash shape

The general modifiers can all be stacked, so they should remain bitfields. The mesh choices are exclusive, so they can be combined into a selector that uses several bits.

So, I'll probably reserve 3 bits for the mesh selector (allowing up to 8 total mesh choices) and maybe 5 bits for the general modifiers (allowing for 2 more bits to be used later).

sofar commented:
Pushed an update addressing these 2 things:

- using a custom CPT2_MESHOPTIONS paramtype2
- don't stretch pixels by 1.4x1, instead scale everything 1.4x1.4 (meaning people who want their plants to stay within 1x1x1 need to make textures 23x16 large (roughly)

sofar commented:
Added docs as well.

paramat commented:
How might this affect render time now that plantlike is a mesh not a simple 2 texture drawtype? Grassland can be very thick and this already has a significant load on rendering.

sofar commented:
Well, since nothing is actually using this right now there would be zero impact.

I'll see if I can do some drawtime tests with a few thousand of the 3 and 4-faced versions of grass, though.

kilbith commented:
> Grassland can be very thick and this already has a significant load on rendering.

Snow nodeboxes are much worse and you still maintain them in place...

est31 commented:
Leaves are very bad too, with your logic we should not have any jungle forests.

sofar commented:
leaves have 3 faces (visible), not two (with culling enabled). But leaves can be placed in several layers high, whereas a grass cover is only ever 1 layer thick. So just going by the math, grass with 4 faces will never be as bad as thick forests, by a margin.

e.g. a jungle tree has a 5x5 footprint, and contains easily 20 or so leaf blocks, let's just say 25. At 3 faces drawn, that's 75 faces per 5x5 foot print. So this would be as many faces as a 100% grass cover (if 3-faced), whereas grass likely never makes it to 50% cover.

Just putting down 1 boat is worse than putting down 500 3-faced grass plants.

sofar commented:
random performance numbers:

- about 8000 grass nodes:
normal: drawtime = 7 with 2.7% jitter
* and # shapes: 7 with 2.7% jitter
* larger mesh: 7 with 2.7% jitter
etc.

This isn't gonna make any significant difference. I'd have to do a million triangles I think.

sofar commented:
![](http://i.imgur.com/h46sOrG.jpg)

1170000 faces, 2.34 million triangles. Waving shader enabled. Drawtime = 9.

celeron55 commented:
It's practically the same amount of geometry; it's not going to make a difference. There are not even different textures.

Performance problems in the mesh generator are not likely either, given the simplicity of all this.

paramat commented:
:)

Ryan-Nolan commented:
I cant wait for this. :)

sofar commented:
I have one more mesh idea I want to add to this.

sofar commented:
I added one more plant mesh, a `#` shaped mesh where the top edges are "leaning" outwards, like a cone shape a bit.

Added one more mesh option that vertically makes each face slightly placed lower. This reduces symmetry and makes plant height slightly (0.125) varied. This is very subtle but very pleasing with simple and regular textures, and fixes symmetry issues people have noted before.

I'm considering this no longer WIP, so removed the label.

sofar commented:
This is the newly added plant mesh:

![](http://i.imgur.com/ULqeLf3.png)

paramat commented:
^ Love that new mesh, it's one i had been thinkiing of.

I disagree with having 'bit 3 (0x8) causes the plant to be randomly offset in the x,z plane'

'bit 4 (0x10) causes the plant mesh to grow by sqrt(2),' surely this can already be done with 'visual scale'?

sofar commented:
@paramat visual scale is an always-on effect that can't be removed or adjusted. By putting it in this way it can be toggled on/off and mods can create more variation in the same map.

paramat commented:
I see now, ok.

paramat commented:
Thought i'd mention that plantlike is rotated by 44 and 46 degrees to avoid texture fighting due to overlap with junglegrass or any texture that extends outside the node. Some of your large variations may have texture fighting when randomised placement is not used, perhaps you need to offset all angles by 1 degree?

sofar commented:
I'll tweak the angles where relevant, expect an update in a day or two.

tobyplowy commented:
@sofar are you still working on this ( it seems to be dead) I'm really hiped for this

paramat commented:
Generally i do like this and am keen for it to be added in some form. The parts i oppose are:

> bit 3 (0x8) causes the plant to be randomly offset in the x,z
plane.

> Added one more mesh option that vertically makes each face slightly placed lower. This reduces symmetry and makes plant height slightly (0.125) varied.

These fine randomisations are i feel unnecessary and out of character, and make the concept disproportionately complex. So overall i'm neutral, but i expect other devs might support.

C1ffisme commented:
@paramat Not all of us like perfect, organized rows and columns of plants in our virtual backyards. Sometimes a bit of imperfection is good too.

tobyplowy commented:
@C1ffisme :D that's actually quite true! Also virtual backyard XD

paramat commented:
How about perfect, organized rows and columns of cubic nodes? =} Having a problem with alignment and repetition in a voxel game is ridiculous, to be consistent we would then offset all cubic nodes randomly.

tobyplowy commented:
Randomly aligned grass looks more realistic and natural

paramat commented:
Randomly aligned cubic nodes look more realistic and natural :]
But that's not a voxel game.

tobyplowy commented:
@paramat not Percy, Minecraft does this as well (that's probably the reason why you don't want it)

paramat commented:
No, it's because it's a bad idea.

ExeterDad commented:
I like this bad idea. "A bunch". 
Which naturally is not aligned.

tobyplowy commented:
We should make a straw poll ( and choose if you want this or Not) it's just an idea

MarkuBu commented:
If there is no technical reason this should be added. I also wait for this feature since month

tobyplowy commented:
@paramat What do you mean by bad idea ( maybe it could be a setting like the waving grass) then we're both happy

sfan5 commented:
:+1:

C1ffisme commented:
@paramat 
> How about perfect, organized rows and columns of cubic nodes?

Actually, there are voxel games which have you digging clumps of dirt and stone, rather than blocks, but that's probably not where we plan to go with Minetest.

The truth is, it's all just aesthetics at this point. You could go for the full randomized world with clumps of stuff and randomized plants, or the perfect rows and columns world. We could try and go for a middle ground though.

You also have to think that someone may make a subgame where the blocks are meshnodes shaped like clumps, so that randomized plants make sense.

MarkuBu commented:
Those who want this feature turn it on, those who don't want it turn it off. Where is the problem?

But if this feature is not available at all those who want's it have lost

I know this does not count, but :+1:

Ferk commented:
Even if this was merged, it actually needs game support to really be used anyway. Even if the plants had this meshoptions paramtype I guess it would require some changes in register_decoration or at the very least in the on_construct code for the plants to use it when generated.

Eeach mod/game could easily decide what values from the meshoptions to use. A bitmask could be added for register decoration to randomize param2, for example. Maybe it's useful for other uses as well.

paramat commented:
Yeah i'm not too concerned about this being added for mod use, the complexity is in using it in mapgen and how to implement it in decoration placing. This may be only be partially implemented in core mapgen. I agree that certain plantlike nodes should have particular drawtypes used (defined by the nodedef 'on place param2' value), and these should be preserved in mapgen, but i'm not keen on going any further than that and supporting the random sizes or random offsets. problem is the expectation that all the new features will be supported by core mapgen, don't get your hopes up for that.

sofar commented:
@paramat I currently would only expect maybe `farming` to use the new plant shapes, really (change wheat shape maybe?). The rest of the options are for mods and subgames.

paramat commented:
Ok cool perhaps we'll go with a partial support in mapgen then, depending on how difficult iturns out to be.
Is this PR ready for merge?

sofar commented:
Let me check the angles don't need to be adjusted slightly to prevent Z-fighting artifacts. Other than that it's good to go.

sofar commented:
So I've cautiously added a 1-degree rotation to all the versions to avoid Z-fighting artifacts. Rebased and re-tested. I think this is good to go now.

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 
 			float s = BS / 2 * f.visual_scale;
+			// add sqrt(2) visual scale
+			if ((f.param_type_2 == CPT2_MESHOPTIONS) && ((n.param2 & 0x10) != 0))
```
Unnecessary brackets?
if (f.param_type_2 == CPT2_MESHOPTIONS && (n.param2 & 0x10) != 0)

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+
+			float random_offset_X = .0;
+			float random_offset_Z = .0;
+			if ((f.param_type_2 == CPT2_MESHOPTIONS) && ((n.param2 & 0x8) != 0)) {
```
As above.

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 					video::S3DVertex(-s,-BS/2, 0, 0,0,0, c, 0,1),
 					video::S3DVertex( s,-BS/2, 0, 0,0,0, c, 1,1),
 					video::S3DVertex( s,-BS/2 + s*2,0, 0,0,0, c, 1,0),
 					video::S3DVertex(-s,-BS/2 + s*2,0, 0,0,0, c, 0,0),
```
4 lines above: spaces around binary operators

sofar replied:
I'm not sure I understand the comment here. Binary operators are things like `&` and `|`, and I don't see those. If you're commenting on lines 1119-1122, then I didn't modify those, and I don't think there is any requirement to fix up style for adjacent sections (it's not prohibited, but requiring it would be illogical as well).

paramat replied:
I mean spaces around / and *.
Just seen spaces after commas are needed too.
I know these are outside your PR but it seems to me that fixing up code style within your own block of changes is a good idea.

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+						case 0:
+							for (u16 i = 0; i < 4; i++) {
+								vertices[i].Pos.rotateXZBy(1);
+								vertices[i].Pos.Z += BS/4;
```
Spaces
BS / 4

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+						case 1:
+							for (u16 i = 0; i < 4; i++) {
+								vertices[i].Pos.rotateXZBy(91);
+								vertices[i].Pos.X += BS/4;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+						case 2:
+							for (u16 i = 0; i < 4; i++) {
+								vertices[i].Pos.rotateXZBy(181);
+								vertices[i].Pos.Z -= BS/4;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+						case 3:
+							for (u16 i = 0; i < 4; i++) {
+								vertices[i].Pos.rotateXZBy(271);
+								vertices[i].Pos.X -= BS/4;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+						switch (j) {
+						case 0:
+							for (u16 i = 2; i < 4; i++)
+								vertices[i].Pos.Z -= BS/2;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+							break;
+						case 1:
+							for (u16 i = 2; i < 4; i++)
+								vertices[i].Pos.Z -= BS/2;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+							break;
+						case 2:
+							for (u16 i = 2; i < 4; i++)
+								vertices[i].Pos.Z -= BS/2;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+							break;
+						case 3:
+							for (u16 i = 2; i < 4; i++)
+								vertices[i].Pos.Z -= BS/2;
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
-				{
+				for (int i = 0; i < 4; i++) {
 					vertices[i].Pos *= f.visual_scale;
 					vertices[i].Pos.Y += BS/2 * (f.visual_scale - 1);
```
and here

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 					vertices[i].Pos.Y += BS/2 * (f.visual_scale - 1);
 					vertices[i].Pos += intToFloat(p, BS);
+					// move to a random spot to avoid moire
+					if ((f.param_type_2 == CPT2_MESHOPTIONS) && ((n.param2 & 0x8) != 0)) {
```
Unnecessary brackets as above

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+						vertices[i].Pos.Z += random_offset_Z;
+					}
+					// randomly move each face up/down
+					if ((f.param_type_2 == CPT2_MESHOPTIONS) && ((n.param2 & 0x20) != 0)) {
```
As above

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+
+				// stop adding faces for meshes with less than 4 faces
+				if (f.param_type_2 == CPT2_MESHOPTIONS) {
+					if (((p2mesh == 0) || (p2mesh == 1)) && (j == 1))
```
if ((p2mesh == 0 || p2mesh == 1) && j == 1)

paramat commented on src/content_mapblock.cpp:
```diff
@@ -1104,45 +1105,174 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
+				if (f.param_type_2 == CPT2_MESHOPTIONS) {
+					if (((p2mesh == 0) || (p2mesh == 1)) && (j == 1))
+						break;
+					else if ((p2mesh == 2) && (j == 2))
```
else if (p2mesh == 2 && j == 2)

paramat commented on src/nodedef.cpp:
```diff
@@ -387,7 +387,10 @@ void ContentFeatures::serialize(std::ostream &os, u16 protocol_version) const
 	writeU8(os, post_effect_color.getBlue());
 	writeU8(os, param_type);
-	writeU8(os, param_type_2);
+	if ((protocol_version < 28) && (param_type_2 == CPT2_MESHOPTIONS))
```
if (protocol_version < 28 && param_type_2 == CPT2_MESHOPTIONS)

paramat commented:
Line comments, otherwise +1.

sofar commented:
I disagree with the parenthesis style comment. Removing them is IMHO bad style, and will lead to compiler warnings if there are cases where multiple conditions are combined (`cond && cond || cond`).

The goal of the parentheses here is to prevent left ordering of `&&` and `||` by accident, and to keep the code human readable without having to remember ordering.

FWICS, nether the minetest style guide nor the linux kernel style has issues with explicit parentheses, which suggests that this is at minimum permitted.

I'll fix the `BS/4` space comment, though.

sofar commented:
Minor spacing adjustment only in last revision.

paramat commented:
Ok.

paramat commented:
There's a missed BS / 2 at line 1248.

So, what's the harm in
if (protocol_version < 28 && param_type_2 == CPT2_MESHOPTIONS)
?
Seems obvious that doesn't need inner brackets.

paramat commented:
else if ((p2mesh == 2) && (j == 2))
Inner braackets are obviously not needed here.

paramat commented:
```if ((p2mesh == 0 || p2mesh == 1) && j == 1)```
I can't see harm in this either, the oder of || and && is clearly defined?

paramat commented:
> will lead to compiler warnings if there are cases where multiple conditions are combined (cond && cond || cond)

That doesn't happen in any of my suggestions.

paramat commented:
```else if ((p2mesh == 2) && (j == 2))```
Oh because it might try to evaluate '2 && j' first?

paramat commented:
Brackets are fine then.
Other stuff can be cleaned up later.

paramat commented:
e58a55aa82dfc66325a694dcc3519d3c0f3388a6

tchncs commented:
Hmm looks awesome, but, is there somewhere an example how to add it to a mod?
I am afraid that i don't understand it without docs. :(
(Docs are done on the checklist but i found nothing in the dev wiki)

paramat commented:
Dev wiki is hopelessly out of date most of the time.

MarkuBu commented:
Is this update available now? I have installed the latest build from launchpad daily build ppa (3 days old). I added 		

paramtype2 == "8" 

to minetest.register_node("default:grass_", but I don't see any effect. Or did I something wrong?

paramat commented:
That's param2 type, not param2 value. Type needs to be 'meshoptions', then use 'place_param2 = 8' to change on player place, howeveer mapgen does not support these mesh options (but might partly do so in future).

paramat commented:
``` paramtype2 == "meshoptions"```
Oops :}

sofar commented:
place_param2 is documented elsewhere:

https://github.com/minetest/minetest/blob/e10fee00011f6c1ef8ee5b884adb11013954a1c9/doc/lua_api.txt#L3612

MarkuBu commented:
Still don't see any effect. Examples would be nice

tchncs commented:
Same here :(
Also it would be reeeally useful to make the mapgen able to use it. ;)

MarkuBu commented:
Sorry, it works now. 

Mapgen support really would be awesome. Make an option to disable the support to minetest.conf

tchncs commented:
@MarkuBu it would be awesome if you share it with us since i am sure there are more people who try to do this and they would be happy to see how it works. :)

Anyway, i tried:

```
    paramtype2 = meshoptions,
    place_param2 = 2,
```

Is this wrong?

paramat commented:
What i'm thinking is that mapgen would apply the param2 set in node definition 'place_param2', as these mesh options will be useful for creating a variety of plant forms.

sofar commented:
decorations can specify `param2` already, which may just be enough already?

sofar commented:
BTW, `place_param2` is meant for "player placement". A mapgen specific option may be needed to distinguish between player-placed and mapgen-placed.

paramat commented:
Decoration placement can't set param2.
Yeah the def name 'place_' may not be suitably named for mapgen placement, but seems the obvious one to use.

MarkuBu commented:
> paramtype2

The name is confusing. Shouldn't it be param2_type?

sofar commented:
@MarkuBu changing that would break many mods, so it is left this way.
